torchtune/modules/early_exit_loss.py

```diff
@@ -0,0 +1,334 @@
+    logits_early = logits_early.contiguous()
+    # Shape: [e*b*s]
+    labels_repeated = labels.repeat(e, 1).reshape(-1)
+    # Compute early losses: Shape: [e*b*s]
+    losses_early = batch_loss_fn(logits_early, labels_repeated)
+    # Shape: [e, b*s]
+    losses_early = losses_early.view(e, -1)
+    # Shape: [e]
+    s_unpadded = (labels != loss_fn.ignore_index).sum()
+    losses_early = losses_early.float().sum(-1) / s_unpadded
+    # Shape: [e]
+    losses_scales = layer_ids_to_loss_scales(
+        torch.Tensor(hidden_layer_ids).to(losses_early),
+        len(model.layers),
+        loss_scale_type,
+        e_scale,
+    )
+
+    return torch.sum(losses_scales * losses_early)
+
+
+def layer_ids_to_loss_scales(
+    layer_ids: torch.Tensor,
+    n_layers: int,
+    loss_scale_type: LossScaleType,
+    e_scale: float,
+) -> torch.Tensor:
+    """
+    Compute the loss scales for a given set of layer IDs and loss scale type.
+    This function takes in a list of layer IDs, the total number of layers,
+    a loss scale type, and an early exit scaling factor. It computes the loss
+    scales based on the specified loss scale type and then normalizes them to
+    ensure that their sum is 1.0.
+    Args:
+        layer_ids (torch.Tensor): A tensor of layer IDs.
+        n_layers (int): The total number of layers.
+        loss_scale_type (LossScaleType): The type of loss scaling to use.
+        e_scale (float): An early exit scaling factor.
+    Returns:
+        torch.Tensor: The computed loss scales.
+    Raises:
+        ValueError: If the provided loss scale type is not supported.
+        AssertionError: If the sum of the loss scales is not close to 1.0.
+    Example:
+        >>> layer_ids = [0, 1, 2]
+        >>> n_layers = 3
+        >>> loss_scale_type = LossScaleType.SUM_L
+        >>> e_scale = 1.0
+        >>> loss_scales = layer_ids_to_loss_scales(layer_ids, n_layers, loss_scale_type, e_scale)
+    """
+    if loss_scale_type == LossScaleType.ONE:
+        loss_scales = torch.ones(len(layer_ids), device=layer_ids.device)
+    elif loss_scale_type == LossScaleType.L:
+        loss_scales = torch.Tensor(layer_ids + 1)
+    elif loss_scale_type == LossScaleType.SUM_L:
+        loss_scales = torch.cumsum(layer_ids + 1, dim=0)
+    elif loss_scale_type == LossScaleType.SQRT_L:
+        loss_scales = torch.sqrt(layer_ids + 1)
+    elif loss_scale_type == LossScaleType.INV_L:
+        loss_scales = 1.0 / (layer_ids + 1)
+    elif loss_scale_type == LossScaleType.INV_SQRT_L:
+        loss_scales = torch.reciprocal(torch.sqrt(layer_ids + 1))
+    else:
+        raise ValueError(f"Unsupported loss_scale type {loss_scale_type}")
+
+    loss_scales = loss_scales * torch.where(layer_ids < n_layers - 1, e_scale, 1.0)
+    # normalize loss scales to ensure that their sum is 1.0
+    loss_scales = loss_scales / torch.sum(loss_scales)
+    assert torch.isclose(torch.sum(loss_scales), torch.Tensor([1.0]).to(loss_scales))
```

Comment:
nit but I don't think this should be needed given the preceding line (are you worried about some sort of numerical issue?)

Reply:
Seems like the assert is still here?

Reply:
Oh Sorry. I made sure I removed the assertion in 2675b4c3fbe5ca4a9f0ea0ace4e9a3f485b87cd4